### PR TITLE
consolidate curl handle creation and add on-the-fly configurability

### LINKED ISSUE #253: Client session hangs indefinitely on long-running query
Client user on a Domino server is facing an indefinite hang for a long-running query. Here is a simple repro:

```R
system.time({scidb::iquery(con$db, "build(<val:int64>[i=0:0], sleep(4))", return=T)})[[3]]
# returns in ~5 secs

system.time({scidb::iquery(con$db, "build(<val:int64>[i=0:0], sleep(412))", return=T)})[[3]]
# hangs indefinitely
```

On the server, side shim logs show that the `/read_bytes` call never came in from the client:

```
Dec 08 17:26:27 ip-172-25-130-11.rdcloud.bms.com shim[32015]: /execute_query
Dec 08 17:26:27 ip-172-25-130-11.rdcloud.bms.com shim[32015]: execute_query[betexb]: execute, scidb[0] 0x7f7a100490b0, scidb[1] 0x7f7a1006b090, query save(show('filter(build(<val:int64>[i=0:0], sleep(412)), true)', 'afl'),'/scidb_data/shim_dir/shim_output_buf_l3fHxb',0,'csv+:l')
Dec 08 17:26:27 ip-172-25-130-11.rdcloud.bms.com shim[32015]: execute_query[betexb]: execute, qid 0.1670520387054913604
Dec 08 17:26:27 ip-172-25-130-11.rdcloud.bms.com shim[32015]: execute_query[betexb]: done
Dec 08 17:26:27 ip-172-25-130-11.rdcloud.bms.com shim[32015]: /read_lines
Dec 08 17:26:27 ip-172-25-130-11.rdcloud.bms.com shim[32015]: read_lines[betexb]: return entire buffer
Dec 08 17:26:27 ip-172-25-130-11.rdcloud.bms.com shim[32015]: /execute_query
Dec 08 17:26:27 ip-172-25-130-11.rdcloud.bms.com shim[32015]: execute_query[betexb]: execute, scidb[0] 0x7f7a100490b0, scidb[1] 0x7f7a1006b090, query aio_save(build(<val:int64>[i=0:0], sleep(412)),paths:('/scidb_data/shim_dir/shim_output_buf_l3fHxb'),instances:(0),format:'(int64 null,int64 )',atts_only:0, result_size_limit:4192)
Dec 08 17:26:27 ip-172-25-130-11.rdcloud.bms.com shim[32015]: execute_query[betexb]: execute, qid 0.1670520387086150733
Dec 08 17:33:19 ip-172-25-130-11.rdcloud.bms.com shim[32015]: execute_query[betexb]: done

# query finished above after 7 min i.e. ~420 sec
# no read bytes call came in from the client

# an hour later, I canceled the query from the client 
# obviously, the query does not exist any more
Dec 08 18:28:36 ip-172-25-130-11.rdcloud.bms.com shim[32015]: /cancel
Dec 08 18:28:36 ip-172-25-130-11.rdcloud.bms.com shim[32015]: cancel[betexb]: execute, qid 0.1670520387086150733, scidb[1] 0x7f7a1006b090, query cancel('0.1670520387086150733')
Dec 08 18:28:36 ip-172-25-130-11.rdcloud.bms.com shim[32015]: cancel[betexb]: result UserQueryException in file: src/query/Query.cpp function: getQueryByID line: 1340 instance: s0-i0 (0)
                                                              Error id: scidb::SCIDB_SE_QPROC::SCIDB_LE_QUERY_NOT_FOUND
                                                              Error description: Query processor error. Query 0.1670520387086150733 not found.
                                                              cancel('0.1670520387086150733')
```

Seems like some kind of "keep-alive" needs to be kept intact between the client on Domino and the scidb server. 

Note that the same long-running query runs OK when the client is on the same server as scidb. 

CC: @CB-93

### PR DESCRIPTION
**Pull request by CB-93**
Fixes #253 

### Problem
* User would query the DB from a Domino Instance
* The query itself takes > 5 minutes to complete
* User observes an indefinite hang in the query (15+ minutes).
* When querying locally, expected behavior observed (successful response and data acquisition via the `/read_bytes` request)

### Solution
* Found that the DB connection was getting timed out, so even though the query does run completely server-side, the client won't know because the response can't get back to the user (which would lead to downstream acquisition via `/read_bytes` e.g.).
* Resolved by locally configuring SciDB-R with TCP keep-alive related options upon curl handle creation

The PR aims to make the solution more flexible and configurable user-side. Meant to be more accommodating to different network configurations or the unique need for certain curl options.

Default behavior would be retained, so no deployments would be directly affected by this PR.

Tests currently failing because of some failure to install R (likely linked to the actions issue where the workflow automatically changed the version of the ubuntu machine it would run? Please correct me otherwise).

#### Side Note
as I let the customer know of this solution, Customer B's UI contractor suggested implementing an `async` approach for longer running queries, while discouraging the forcing open of connections as a long term solution. Integrating this approach now would be potentially tricky/take longer to do (e.g. I wouldn't want to `async` for quick queries, but finding out how big a query would be (proxy for query time) without querying the data itself -- whether by `show` or `summarize` -- would be non-trivial, so the user would have to know beforehand and perhaps set an `async` parameter.) Something to opine on later...

### CODE REVIEW

File: R/internal.R [hunk 1]
```diff
@@ -570,22 +561,15 @@ POST = function(db, data, args=list(), err=TRUE)
     uri = URI(db, "/upload", args)
     uri = oldURLencode(uri)
     uri = gsub("\\+", "%2B", uri, perl=TRUE)
-    h = new_handle()
-    handle_setheaders(h, .list=list(Authorization=digest_auth(db, "POST", uri)))
-    handle_setopt(h, .list=list(ssl_verifyhost=as.integer(getOption("scidb.verifyhost", FALSE)),
-                                ssl_verifypeer=0, post=TRUE, http_version=2, postfieldsize=length(data),
```
Comment (agvandervelde) [inline R/internal.R:576]: `postfieldsize` and `postfields` need special handling here.

File: R/internal.R [hunk 2]
```diff
@@ -539,10 +533,7 @@ SGET = function(db, resource, args=list(), err=TRUE, binary=FALSE)
   uri = URI(db, resource, args)
   uri = oldURLencode(uri)
   uri = gsub("\\+", "%2B", uri, perl=TRUE)
-  h = new_handle()
-  handle_setheaders(h, .list=list(Authorization=digest_auth(db, "GET", uri)))
```
Comment (agvandervelde) [inline R/internal.R:543]: Some of those are POST and others are GET

File: R/internal.R [hunk 3]
```diff
@@ -570,22 +561,15 @@ POST = function(db, data, args=list(), err=TRUE)
     uri = URI(db, "/upload", args)
     uri = oldURLencode(uri)
     uri = gsub("\\+", "%2B", uri, perl=TRUE)
-    h = new_handle()
-    handle_setheaders(h, .list=list(Authorization=digest_auth(db, "POST", uri)))
```
Comment (agvandervelde) [inline R/internal.R:574]: "POST"

File: R/internal.R [hunk 4]
```diff
@@ -1172,3 +1156,108 @@ multipart_post_load <- function(db, session,
 
   return(NULL)
 }
+
+#' Create a Curl Handle to be Used in DB Queries
+#' @keywords internal
+#' @importFrom curl new_handle handle_setheaders handle_setopt
+create_curl_handle <- function(db, uri){
```
Comment (agvandervelde) [inline R/internal.R:1163]: You probably want to add a few arguments here to account for diversity in current calls.

**Review by agvandervelde — CHANGES_REQUESTED** [review]

File: R/internal.R [hunk 5]
```diff
@@ -1172,3 +1158,120 @@ multipart_post_load <- function(db, session,
...
+    if(length(invalid_options) > 0){
+      valid = F # for consistency
+      stop( # message, or stop?
+        glue::glue(
+          "These options in scidb.curl_options aren't valid:
+{paste(invalid_options, collapse=', ')}"
+        )
+      )
+    } else{
+      valid = T
```
Comment (agvandervelde) [inline R/internal.R:1269]: small note on style: `T` -> `TRUE`, `F` -> `FALSE`

File: R/internal.R [hunk 6]
```diff
@@ -1172,3 +1158,120 @@ multipart_post_load <- function(db, session,
...
+  return(
+    list(
+      headers = handle_headers,
+      options = handle_options
+    )
+  )
+}
+
+#' Check if curl options set by user are valid options from curl library
+#' @keywords internal
```
Comment (agvandervelde) [inline R/internal.R:1224]: If we're adding documentation, might as well document `@param`s and `@return`s.

File: R/internal.R [hunk 7]
```diff
@@ -1172,3 +1158,120 @@ multipart_post_load <- function(db, session,
...
+  db, request_type, uri, postdata=NULL
+  ){
+  # Default Behavior
+  # dont' have these get overwritten
+  handle_headers = list(
+    `Authorization`=digest_auth(db, request_type, uri)
+  )
+  handle_options = list(
+    ssl_verifyhost=as.integer(getOption("scidb.verifyhost", FALSE)),
+    ssl_verifypeer=0,
```
Comment (agvandervelde) [inline R/internal.R:1197]: probably make this an option too, with default == 0

**Review by agvandervelde — CHANGES_REQUESTED** [review]

File: R/internal.R [hunk 8]
```diff
@@ -1172,3 +1158,118 @@ multipart_post_load <- function(db, session,
...
+    option_names = names(scidb_curl_options)
+    # check if defaults are getting overwritten/duplicated
+    overwritten_options = intersect(
+      option_names,
+      reserved_options
+    )
+    if(length(overwritten_options) > 0){
+      valid = FALSE
+      warning(
+        glue::glue(
```
Comment (agvandervelde) [inline R/internal.R:1259]: @CB-93 it might be better to turn this into a `paste0`, to avoid adding `glue` as a dependency.

**Review by agvandervelde — APPROVED** [review]

File: R/internal.R [hunk 9]
```diff
@@ -1203,10 +1203,10 @@ formulate_curl_handle_and_opts = function(
   )
   handle_options = list(
     ssl_verifyhost=as.integer(getOption("scidb.verifyhost", FALSE)),
-    ssl_verifypeer=0,
+    ssl_verifypeer=as.integer(getOption("scidb.verifypeer", FALSE)),
```
Comment (agvandervelde) [inline R/internal.R:1206]: Oh, the default here needs to be `0`, not `FALSE`.

**Review by agvandervelde — CHANGES_REQUESTED** [review]

File: R/internal.R (hunk 9 above)
Comment (agvandervelde) [reply R/internal.R]: ok, makes sense.

**Review by agvandervelde — APPROVED** [review]

**Review by apoliakov — APPROVED** [review]
Hey guys, this looks good. Thanks for the work!

**Review by ksens — CHANGES_REQUESTED** [review]
LGTM... Tests seem to be passing after merging master to this branch. 

Only change requested - update DESCRIPTION and NEWS files with details of a bumped package version number.

**Pull request merged**